lib/ansible/utils/collection_loader/_collection_finder.py

```diff
@@ -211,19 +212,22 @@ def find_module(self, fullname, path=None):
         try:
             if toplevel_pkg == 'ansible':
                 # something under the ansible package, delegate to our internal loader in case of redirections
-                return _AnsibleInternalRedirectLoader(fullname=fullname, path_list=path)
-            if part_count == 1:
-                return _AnsibleCollectionRootPkgLoader(fullname=fullname, path_list=path)
-            if part_count == 2:  # ns pkg eg, ansible_collections, ansible_collections.somens
-                return _AnsibleCollectionNSPkgLoader(fullname=fullname, path_list=path)
+                loader = _AnsibleInternalRedirectLoader(fullname=fullname, path_list=path)
+            elif part_count == 1:
+                loader = _AnsibleCollectionRootPkgLoader(fullname=fullname, path_list=path)
+            elif part_count == 2:  # ns pkg eg, ansible_collections, ansible_collections.somens
+                loader = _AnsibleCollectionNSPkgLoader(fullname=fullname, path_list=path)
             elif part_count == 3:  # collection pkg eg, ansible_collections.somens.somecoll
-                return _AnsibleCollectionPkgLoader(fullname=fullname, path_list=path)
-            # anything below the collection
-            return _AnsibleCollectionLoader(fullname=fullname, path_list=path)
+                loader = _AnsibleCollectionPkgLoader(fullname=fullname, path_list=path)
+            else:
+                # anything below the collection
+                loader = _AnsibleCollectionLoader(fullname=fullname, path_list=path)
         except ImportError:
             # TODO: log attempt to load context
             return None
 
+        return importlib.util.spec_from_loader(fullname, loader)
```

Comment:
It's likely not necessary for our use, but you could pass `origin` here also.

```
origin=loader.get_filename(loader._fullname)
```

One thing to note is that `_AnsibleInternalRedirectLoader` doesn't implement `get_filename` so could just be added to return `None`.